### PR TITLE
[SPARK-14955] [SQL] avoid stride value equals to zero

### PR DESCRIPTION
**Pull request by bomeng**
## What changes were proposed in this pull request?

In the `columnPartition()` method of `JDBCRelation`, stride is used for calculating the increment. But in some cases, this value could be zero, for example, lowerBound=0, upperBound=7, numOfPartition=8, which put all the data into one partition (last partition).

This fix will try to make stride calculation more robust. I have also added require() to validate the input. `equals()` is added for overriding the parents `equals()` method, together with `hashCode()`.

I have also fixed some text style, make keywords all uppercase.
## How was this patch tested?

New test cases were added to `JDBCSuite`.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 1]
```diff
@@ -40,6 +40,16 @@ import org.apache.spark.unsafe.types.UTF8String
  */
 private[sql] case class JDBCPartition(whereClause: String, idx: Int) extends Partition {
   override def index: Int = idx
+
+  override def hashCode(): Int = idx
+
+  override def equals(other: Any): Boolean = {
+    other match {
+      case partition: JDBCPartition =>
+        if (whereClause == partition.whereClause && idx == partition.idx) true else false
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: `if (x) true else false` is reducible to just `x`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala [hunk 2]
```diff
@@ -54,15 +54,22 @@ private[sql] object JDBCRelation {
...
+    require(lower < upper, "lower bound must be less than upper bound")
+    require(numPartitions > 0, "number of partition must be great than zero")
+
     if (numPartitions == 1) return Array[Partition](JDBCPartition(null, 0))
-    // Overflow and silliness can happen if you subtract then divide.
-    // Here we get a little roundoff, but that's (hopefully) OK.
-    val stride: Long = (partitioning.upperBound / numPartitions
-                      - partitioning.lowerBound / numPartitions)
+
+    val stride: Long = {
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: `val stride = math.max(1, (upper - lower + 1) / numPartitions)`

Comment (bomeng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRelation.scala]: Cool, I will update it shortly.

**Comment by bomeng** [comment]
@srowen Please review again. Thanks.

**Comment by srowen** [comment]
I don't feel qualified to evaluate the logic change, but the code style itself looks OK

**Comment by bomeng** [comment]
@andrewor14 Hey Andrew, could you please review this one as well?

**Comment by bomeng** [comment]
close this pr provided it was fixed by another pr.

**Comment by HyukjinKwon** [comment]
@bomeng, do you mind if I ask to point out the PR fixing this issue if possible please?

**Pull request closed without merging**